docs/rules/no-controllers.md

```diff
@@ -0,0 +1,32 @@
+# no-controllers
+
+Some people may prefer to avoid the use of controllers in their applications, typically in favor of components which can be more portable and easier to test.
+
+This rule disallows controller usage, except when the controller is used to define `queryParams`.
+
+## Rule Detail
+
+Examples of **incorrect** code for this rule:
+
+```js
+import Controller from '@ember/controller';
+export default class ArticlesController extends Controller {
+  // Controller disallowed since it doesn't use `queryParams`.
+  ...
+}
+```
+
+Examples of **correct** code for this rule:
+
+```js
+import Controller from '@ember/controller';
+export default class ArticlesController extends Controller {
+  queryParams = ['category']; // Controller allowed for defining `queryParams`.
+  @tracked category = null;
+  ...
+}
+```
```

Comment:
Would the rule prevent code other than queryParams and the category property?

Reply:
Currently no. I'm enforcing the following simplified rule: If `queryParams` is defined inside the controller, then the controller and any other code in it is allowed. If `queryParams` is not defined inside the controller, then the controller is not allowed.

Why? Because it's non-trivial to determine if a particular line of code/function/property/observer/action handler/service injection/etc inside a controller is involved in or needed for reading or modifying a query parameter value.

For example, all the code below is needed for managing a query parameter value, but it would be very difficult for the lint rule to determine that.

```js
export default Controller.extend({
  queryParams: ['sortType'],
  sortType: null,

  hasDefaultSortType: equal('model.sortType', DEFAULT_SORT_TYPE),

  sortObserver: observer('hasDefaultSortType', function() {
    if (this.hasDefaultSortType) {
      this.set('sortType', null);
    } else {
      this.set('sortType', this.model.sortType);
    }
  }),
});
```

Reply:
makes sense. what happens if `queryParams = []` is defined?

Reply:
If you're asking if the rule handles native class syntax, then yes, it will allow a native class controller as long as it has `queryParams = [...]`.